README.md
# Benchmark::Trend

[![Gem Version](https://badge.fury.io/rb/benchmark-trend.svg)][gem]
[![Build Status](https://secure.travis-ci.org/piotrmurach/benchmark-trend.svg?branch=master)][travis]
[![Build status](https://ci.appveyor.com/api/projects/status/798apneaa8ixg5dk?svg=true)][appveyor]
[![Maintainability](https://api.codeclimate.com/v1/badges/782faa4a8a4662c86792/maintainability)][codeclimate]
[![Coverage Status](https://coveralls.io/repos/github/piotrmurach/benchmark-trend/badge.svg?branch=master)][coverage]
[![Inline docs](http://inch-ci.org/github/piotrmurach/benchmark-trend.svg?branch=master)][inchpages]

[gem]: http://badge.fury.io/rb/benchmark-trend
[travis]: http://travis-ci.org/piotrmurach/benchmark-trend
[appveyor]: https://ci.appveyor.com/project/piotrmurach/benchmark-trend
[codeclimate]: https://codeclimate.com/github/piotrmurach/benchmark-trend/maintainability
[coverage]: https://coveralls.io/github/piotrmurach/benchmark-trend?branch=master
[inchpages]: http://inch-ci.org/github/piotrmurach/benchmark-trend

> Measure performance trends of Ruby code based on the input size distribution.

**Benchmark::Trend** will help you estimate the computational complexity of Ruby code by running it on inputs increasing in size, measuring their execution times, and then fitting these observations into a model that best predicts how a given Ruby code will scale as a function of growing workload.

## Why?

Tests provide safety net that ensures your code works correctly. What you don't know is how fast your code is! How does it scale with different input sizes? Your code may have computational complexity that doesn't scale with large workloads. It would be good to know before your application goes into production, wouldn't it?

**Benchmark::Trend** will allow you to uncover performance bugs or confirm that a Ruby code performance scales as expected.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'benchmark-trend'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install benchmark-trend

## Contents

* [1. Usage](#1-usage)
* [2. API](#2-api)
  * [2.1 range](#21-range)
  * [2.2 infer_trend](#22-infer_trend)
    * [2.2.1 repeat](#221-repeat)
  * [2.3 fit](#23-fit)
  * [2.4 fit_at](#24-fit_at)
* [3. Examples](#3-examples)
  * [3.1 Ruby array max](#31-ruby-array-max)

## 1. Usage

Let's assume we would like to find out behaviour of a Fibonacci algorithm:

```ruby
def fibonacci(n)
  n < 2 ? n : fibonacci(n - 1) + fibonacci(n - 2)
end
```

To measure the actual complexity of above function, we will use `infer_trend` method and pass it as a first argument an array of integer sizes and a block to execute the method:

```ruby
numbers = Benchmark::Trend.range(1, 28, ratio: 2)

trend, trends = Benchmark::Trend.infer_trend(numbers) do |n, i|
  fibonacci(n)
end
```

The return type will provide a best trend name:

```ruby
print trend
# => exponential
```

and a Hash of all the trend data:

```ruby
print trends
# =>
# {:exponential=>
#   {:trend=>"1.38 * 0.00^x",
#    :slope=>1.382889711685203,
#    :intercept=>3.822775903539121e-06,
#    :residual=>0.9052392775178072},
#  :power=>
#   {:trend=>"0.00 * x^2.11",
#    :slope=>2.4911044372815657e-06,
#    :intercept=>2.1138475434240918,
#    :residual=>0.5623418036957115},
#  :linear=>
#   {:trend=>"0.00 + -0.01*x",
#    :slope=>0.0028434594496586007,
#    :intercept=>-0.01370769842204958,
#    :residual=>0.7290365425188893},
#  :logarithmic=>
#   {:trend=>"0.02 + -0.02*ln(x)",
#    :slope=>0.01738674709454521,
#    :intercept=>-0.015489004560847924,
#    :residual=>0.3982368125757882}}
```

You can see information for the best trend by passing name into trends hash:

```ruby
print trends[trend]
# =>
# {:trend=>"1.38 * 0.00^x",
#  :slope=>1.382889711685203,
#  :intercept=>3.822775903539121e-06,
#  :residual=>0.9052392775178072},
```

## 2. API

### 2.1 range

To generate a range of values for testing code fitness use the `range` method. It will generate a geometric sequence of numbers, where intermediate values are powers of range multiplier, by default 8:

```ruby
Benchmark::Trend.range(8, 8 << 10)
# => [8, 64, 512, 4096, 8192]
```

You can change the default sequence power by using `:ratio` keyword:

```ruby
Benchmark::Trend.range(8, 8 << 10, ratio: 2)
# => [8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192]
```

### 2.2 infer_trend

To calculate an asymptotic behaviour of Ruby code by inferring its computational complexity use `infer_trend`. This method takes as an argument an array of inputs which can be generated using [range](#21-range). The code to measure needs to be provided inside a block. Two parameters are always yielded to a block, first, the actual data input and second the current index matching the input.

For example, let's assume you would like to find out asymptotic behaviour of a Fibonacci algorithm:

```ruby
def fibonacci(n)
  n < 2 ? n : fibonacci(n - 1) + fibonacci(n - 2)
end
```

You could start by generating a range of inputs in powers of 2:

```ruby
numbers = Benchmark::Trend.range(1, 32, ratio: 2)
# => [1, 2, 4, 8, 16, 32]
```

Then measure the performance of the Fibonacci algorithm for each of the data points and fit the observations into a model to predict behaviour as a function of input size:

```ruby
trend, trends = Benchmark::Trend.infer_trend(numbers) do |n, i|
  fibonacci(n)
end
```

The return includes the best fit name:

```ruby
print trend
# => exponential
```

And a Hash of all measurements:

```ruby
print trends
# =>
# {:exponential=>
#   {:trend=>"1.38 * 0.00^x",
#    :slope=>1.382889711685203,
#    :intercept=>3.822775903539121e-06,
#    :residual=>0.9052392775178072},
#  :power=>
#   {:trend=>"0.00 * x^2.11",
#    :slope=>2.4911044372815657e-06,
#    :intercept=>2.1138475434240918,
#    :residual=>0.5623418036957115},
#  :linear=>
#   {:trend=>"0.00 + -0.01*x",
#    :slope=>0.0028434594496586007,
#    :intercept=>-0.01370769842204958,
#    :residual=>0.7290365425188893},
#  :logarithmic=>
#   {:trend=>"0.02 + -0.02*ln(x)",
#    :slope=>0.01738674709454521,
#    :intercept=>-0.015489004560847924,
#    :residual=>0.3982368125757882}}
```

In order to retrieve trend data for the best fit do:

```ruby
print trends[trend]
# =>
# {:trend=>"1.38 * 0.00^x",
#  :slope=>1.382889711685203,
#  :intercept=>3.822775903539121e-06,
#  :residual=>0.9052392775178072}
```

### 2.2.1 repeat

To increase stability of you tests consider repeating all time execution measurements using `:repeat` keyword.

Start by generating a range of inputs for your algorithm:

```ruby
numbers = Benchmark::Trend.range(1, 32, ratio: 2)
# => [1, 2, 4, 8, 16, 32]
```

and then run your algorithm for each input repeating measurements `100` times:

```ruby
Benchmark::Trend.infer_trend(numbers, repeat: 100) { |n, i| ... }
```

### 2.3 fit

Use `fit` method if you wish to fit arbitrary data into a model with a slope and intercept parameters that minimize the error.

For example, given a set of data points that exhibit linear behaviour:

```ruby
xs = [1, 2, 3, 4, 5]
ys = xs.map { |x| 3.0 * x + 1.0 }
```

Fit the data into a model:

```ruby
slope, intercept, error = Benchmark::Trend.fit(xs, ys)
```

And printing the values we get confirmation of the linear behaviour of the data points:

```ruby
print slope
# => 3.0
print intercept
# => 1.0
print error
# => 1.0
```

### 2.4 fit_at

If you are interested how a model scales for a given input use `fit_at`. This method expects that there is a fit model generated using [infer_trend](#22-infer_trend).

For example, measuring Fibonacci recursive algorithm we have the following results:

```ruby
# =>
# {:trend=>"1.38 * 0.00^x",
#  :slope=>1.382889711685203,
#  :intercept=>3.822775903539121e-06,
#  :residual=>0.9052392775178072}
```

And checking model at input of `50`:

```ruby
Benchamrk::Trend.fit_at(:exponential, slope: 1.382889711685203, intercept: 3.822775903539121e-06, n: 50)
# => 41.8558455915123
```

We can see that Fibonacci with just a number 50 will take around 42 seconds to get the result!

How about Fibonacci with 100 as an input?

```ruby
Benchamrk::Trend.fit_at(:exponential, slope: 1.382889711685203, intercept: 3.822775903539121e-06, n: 100)
# => 458282633.9777338
```

This means Fibonacci recursive algorithm will take about 1.45 year to complete!

## 3. Examples

### 3.1 Ruby array max

Suppose you wish to find an asymptotic behaviour of Ruby built Array `max` method.

You could start with generating a [range](#21-range) of inputs:

```ruby
array_sizes = Benchmark::Trend.range(1, 100_000)
# => [1, 8, 64, 512, 4096, 32768, 100000]
```

Next, based on the generated ranges create arrays containing randomly generated integers:

```ruby
number_arrays = array_sizes.map { |n| Array.new(n) { rand(n) } }
```

Then feed this information to infer a trend:

```ruby
trend, trends = Benchmark::Trend.infer_trend(array_sizes) do |n, i|
  number_arrays[i].max
end
```

Unsurprisingly, we discover that Ruby's `max` call scales linearily with the input size:

```ruby
print trend
# => linear
```

We can also see from the residual value that this is a near perfect fit:

```ruby
print trends[trend]
# =>
# {:trend=>"0.00 + 0.00*x",
#  :slope=>5.873536409841244e-09,
#  :intercept=>3.028647045635842e-05,
#  :residual=>0.9986764704492359}
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/benchmark-trend. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Benchmark::Trend project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/piotrmurach/benchmark-trend/blob/master/CODE_OF_CONDUCT.md).

## Copyright

Copyright (c) 2018 Piotr Murach. See LICENSE for further details.


CHANGELOG.md
# Change log

## [v0.4.0] - 2020-03-01

### Added
* Add Clock for monotonic time measuring

### Changed
* Change to remove benchmark requirement
* Change gemspec to add metadata, remove test artefacts

## [v0.3.0] - 2019-04-21

### Changed
* Change to require Ruby >= 2.0.0
* Change to relax development dependencies

## [v0.2.0] - 2018-09-30

### Added
* Add ability to measure monotonic time
* Add ability to repeat measurements to increase stability of execution times

### Changed
* Change to prefer simpler complexity for similar measurements
* Change to use monotonic clock
* Change to differentiate linear vs logarithmic complexity for small values
* Change to differentiate linear vs constant complexity for small values

## Fixed
* Fix fit_power to correctly calculate slope and intercept

## [v0.1.0] - 2018-09-08

* Initial implementation and release

[v0.4.0]: https://github.com/piotrmurach/benchmark-trend/compare/v0.3.0...v0.4.0
[v0.3.0]: https://github.com/piotrmurach/benchmark-trend/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/piotrmurach/benchmark-trend/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/piotrmurach/benchmark-trend/compare/v0.1.0


spec/spec_helper.rb
# frozen_string_literal: true

if ENV["COVERAGE"] || ENV["TRAVIS"]
  require "simplecov"
  require "coveralls"

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
    SimpleCov::Formatter::HTMLFormatter,
    Coveralls::SimpleCov::Formatter
  ]

  SimpleCov.start do
    command_name "spec"
    add_filter "spec"
  end
end

require "bundler/setup"
require "benchmark/trend"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/unit/range_spec.rb
# frozen_string_literal: true

RSpec.describe Benchmark::Trend, "#range" do
  it "creates default range" do
    range = Benchmark::Trend.range(8, 8 << 10)
    expect(range).to eq([8, 64, 512, 4096, 8192])
  end

  it "creates range with 2 multiplier" do
    range = Benchmark::Trend.range(8, 8 << 10, ratio: 2)
    expect(range).to eq([8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192])
  end

  it "checks range start to be valid" do
    expect { 
      Benchmark::Trend.range(-2, 10_000)
    }.to raise_error(ArgumentError, "Range value: -2 needs to be greater than 0")
  end

  it "checks range end to be greater than start" do
    expect { 
      Benchmark::Trend.range(8, 2)
    }.to raise_error(ArgumentError, "Range value: 2 needs to be greater than 8")
  end

  it "checks multiplier to be valid" do
    expect { 
      Benchmark::Trend.range(8, 32, ratio: 1)
    }.to raise_error(ArgumentError, "Range value: 1 needs to be greater than 2")
  end
end


spec/unit/fit_at_spec.rb
# frozen_string_literal: true

RSpec.describe Benchmark::Trend, "#fit_at" do
  it "evalutes logarithmic fit model at a given value" do
    val = Benchmark::Trend.fit_at(:logarithmic, slope: 1.5, intercept: 2, n: 10)

    expect(val).to be_within(0.1).of(5.45)
  end

  it "evalutes linear fit model at a given value" do
    val = Benchmark::Trend.fit_at(:linear, slope: 1.5, intercept: 2, n: 10)

    expect(val).to eq(17)
  end

  it "evalutes power fit model at a given value" do
    val = Benchmark::Trend.fit_at(:power, slope: 1.5, intercept: 2, n: 10)

    expect(val).to be_within(0.1).of(63.24)
  end

  it "evalutes power fit model at a given value" do
    val = Benchmark::Trend.fit_at(:exponential, slope: 1.5, intercept: 2, n: 10)

    expect(val).to be_within(0.1).of(115.33)
  end

  it "doesn't recognise fit model" do
    expect {
      Benchmark::Trend.fit_at(:unknown, slope: 1.5, intercept: 2, n: 10)
    }.to raise_error(ArgumentError, "Unknown fit type: unknown")
  end

  it "doesn't allow incorrect input size" do
    expect {
      Benchmark::Trend.fit_at(:linear, slope: 1.5, intercept: 2, n: -1)
    }.to raise_error(ArgumentError, "Incorrect input size: -1")
  end
end


lib/benchmark/trend.rb
# frozen_string_literal: true

require 'benchmark'

require_relative 'trend/version'

module Benchmark
  module Trend
    # Change module function visiblity to private
    #
    # @api private
    def self.private_module_function(method)
      module_function(method)
      private_class_method(method)
    end

    if defined?(Process::CLOCK_MONOTONIC)
      # Object representing current time
      def time_now
        Process.clock_gettime Process::CLOCK_MONOTONIC
      end
      module_function :time_now
    else
      # Object represeting current time
      def time_now
        Time.now
      end
      module_function :time_now
    end

    # Measure time elapsed with a monotonic clock
    #
    # @public
    def clock_time
      before = time_now
      yield
      after = time_now
      after - before
    end
    module_function :clock_time

    # Generate a range of inputs spaced by powers.
    #
    # The default range is generated in the multiples of 8.
    #
    # @example
    #   Benchmark::Trend.range(8, 8 << 10)
    #   # => [8, 64, 512, 4096, 8192]
    #
    # @param [Integer] start
    # @param [Integer] limit
    # @param [Integer] ratio
    #
    # @api public
    def range(start, limit, ratio: 8)
      check_greater(start, 0)
      check_greater(limit, start)
      check_greater(ratio, 2)

      items = []
      count = start
      items << count
      (limit / ratio).times do
        count *= ratio
        break if count >= limit
        items << count
      end
      items << limit if start != limit
      items
    end
    module_function :range

    # Check if expected value is greater than minimum
    #
    # @param [Numeric] expected
    # @param [Numeric] min
    #
    # @raise [ArgumentError]
    #
    # @api private
    def check_greater(expected, min)
      unless expected >= min
        raise ArgumentError,
              "Range value: #{expected} needs to be greater than #{min}"
      end
    end
    private_module_function :check_greater

    # Gather times for each input against an algorithm
    #
    # @param [Array[Numeric]] data
    #   the data to run measurements for
    #
    # @param [Integer] repeat
    #   nubmer of times work is called to compute execution time
    #
    # @return [Array[Array, Array]]
    #
    # @api public
    def measure_execution_time(data = nil, repeat: 1, &work)
      inputs = data || range(1, 10_000)
      times  = []

      inputs.each_with_index do |input, i|
        GC.start
        measurements = []

        repeat.times do
          measurements << clock_time { work.(input, i) }
        end

        times << measurements.reduce(&:+).to_f / measurements.size
      end
      [inputs, times]
    end
    module_function :measure_execution_time

    # Finds a line of best fit that approximates linear function
    #
    # Function form: y = ax + b
    #
    # @param [Array[Numeric]] xs
    #   the data points along X axis
    #
    # @param [Array[Numeric]] ys
    #   the data points along Y axis
    #
    # @return [Numeric, Numeric, Numeric]
    #   return a slope, b intercept and rr correlation coefficient
    #
    # @api public
    def fit_linear(xs, ys)
      fit(xs, ys)
    end
    module_function :fit_linear

    # Find a line of best fit that approximates logarithmic function
    #
    # Model form: y = a*lnx + b
    #
    # @param [Array[Numeric]] xs
    #   the data points along X axis
    #
    # @param [Array[Numeric]] ys
    #   the data points along Y axis
    #
    # @return [Numeric, Numeric, Numeric]
    #   returns a, b, and rr values
    #
    # @api public
    def fit_logarithmic(xs, ys)
      fit(xs, ys, tran_x: ->(x) { Math.log(x) })
    end
    module_function :fit_logarithmic

    alias fit_log fit_logarithmic
    module_function :fit_log

    # Finds a line of best fit that approxmimates power function
    #
    # Function form: y = bx^a
    #
    # @return [Numeric, Numeric, Numeric]
    #   returns a, b, and rr values
    #
    # @api public
    def fit_power(xs, ys)
      a, b, rr = fit(xs, ys, tran_x: ->(x) { Math.log(x) },
                             tran_y: ->(y) { Math.log(y) })

      [a, Math.exp(b), rr]
    end
    module_function :fit_power

    # Find a line of best fit that approximates exponential function
    #
    # Model form: y = ab^x
    #
    # @return [Numeric, Numeric, Numeric]
    #   returns a, b, and rr values
    #
    # @api public
    def fit_exponential(xs, ys)
      a, b, rr = fit(xs, ys, tran_y: ->(y) { Math.log(y) })

      [Math.exp(a), Math.exp(b), rr]
    end
    module_function :fit_exponential

    alias fit_exp fit_exponential
    module_function :fit_exp

    # Fit the performance measurements to construct a model with
    # slope and intercept parameters that minimize the error.
    #
    # @param [Array[Numeric]] xs
    #   the data points along X axis
    #
    # @param [Array[Numeric]] ys
    #   the data points along Y axis
    #
    # @return [Array[Numeric, Numeric, Numeric]
    #   returns slope, intercept and model's goodness-of-fit
    #
    # @api public
    def fit(xs, ys, tran_x: ->(x) { x }, tran_y: ->(y) { y })
      eps    = (10 ** -10)
      n      = 0
      sum_x  = 0.0
      sum_x2 = 0.0
      sum_y  = 0.0
      sum_y2 = 0.0
      sum_xy = 0.0

      xs.zip(ys).each do |x, y|
        n        += 1
        sum_x    += tran_x.(x)
        sum_y    += tran_y.(y)
        sum_x2   += tran_x.(x) ** 2
        sum_y2   += tran_y.(y) ** 2
        sum_xy   += tran_x.(x) * tran_y.(y)
      end

      txy = n * sum_xy - sum_x * sum_y
      tx  = n * sum_x2 - sum_x ** 2
      ty  = n * sum_y2 - sum_y ** 2

      is_linear = tran_x.(Math::E) * tran_y.(Math::E) == Math::E ** 2

      if tx.abs < eps # no variation in xs
        raise ArgumentError, "No variation in data #{xs}"
      elsif ty.abs < eps && is_linear # no variation in ys - constant fit
        slope = 0
        intercept = sum_y / n
        residual_sq = 1 # doesn't exist
      else
        slope       = txy / tx
        intercept   = (sum_y - slope * sum_x) / n
        residual_sq = (txy ** 2) / (tx * ty)
      end

      [slope, intercept, residual_sq]
    end
    module_function :fit

    # Take a fit and estimate behaviour at input size n
    #
    # @example
    #   fit_at(:power, slope: 1.5, intercept: 2, n: 10)
    #
    # @return
    #   fit model value for input n
    #
    # @api public
    def fit_at(type, slope: nil, intercept: nil, n: nil)
      raise ArgumentError, "Incorrect input size: #{n}" unless n > 0

      case type
      when :logarithmic, :log
        intercept + slope * Math.log(n)
      when :linear
        intercept + slope * n
      when :power
        intercept * (n ** slope)
      when :exponential, :exp
        intercept * (slope ** n)
      else
        raise ArgumentError, "Unknown fit type: #{type}"
      end
    end
    module_function :fit_at

    # A mathematical notation template for a trend type
    #
    # @param [String] type
    #   the fit model type
    #
    # @return [String]
    #   the formatted mathematical function template
    #
    # @api private
    def format_fit(type)
      case type
      when :logarithmic, :log
        "%.2f + %.2f*ln(x)"
      when :linear, :constant
        "%.2f + %.2f*x"
      when :power
        "%.2f * x^%.2f"
      when :exponential, :exp
        "%.2f * %.2f^x"
      else
        raise ArgumentError, "Unknown type: '#{type}'"
      end
    end
    module_function :format_fit

    # the trends to consider
    FIT_TYPES = [:exponential, :power, :linear, :logarithmic]

    # Infer trend from the execution times
    #
    # Fits the executiom times for each range to several fit models.
    #
    # @param [Integer] repeat
    #   nubmer of times work is called to compute execution time
    #
    # @yieldparam work
    #   the block of which the complexity is measured
    #
    # @return [Array[Symbol, Hash]]
    #   the best fitting and all the trends
    #
    # @api public
    def infer_trend(data, repeat: 1, &work)
      ns, times = *measure_execution_time(data, repeat: repeat, &work)
      best_fit = :none
      best_residual = 0
      fitted = {}
      n = ns.size.to_f
      aic = -1.0/0
      best_aic = -1.0/0

      FIT_TYPES.each do |fit|
        a, b, rr = *send(:"fit_#{fit}", ns, times)
        # goodness of model
        aic = n * (Math.log(Math::PI) + 1) + n * Math.log(rr / n)
        if a == 0 && fit == :linear
          fit = :constant
        end
        fitted[fit] = { trend: format_fit(fit) % [a, b],
                        slope: a, intercept: b, residual: rr }
        if rr >= best_residual && aic >= best_aic
          best_residual = rr
          best_fit = fit
          best_aic = aic
        end
      end

      [best_fit, fitted]
    end
    module_function :infer_trend
  end # Trend
end # Benchmark
